feat(core): Create StaticInjector which does not depend on Reflect polyfill

mhevery opened the pull request.

mhevery committed:
feat(core): Create StaticInjector which does not depend on Reflect polyfill.

devoto13 commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,333 @@ export abstract class Injector {
+
+  get<T>(token: Type<T>|InjectionToken<T>, notFoundValue?: T): T;
+  get(token: any, notFoundValue?: any): any;
+  get(token: any, notFoundValue?: any): any {
```
Is duplicate signature intended?

mhevery replied:
Yes, see https://www.typescriptlang.org/docs/handbook/functions.html#overloads

mhevery committed:
perf: switch angular to use StaticInjector instead of ReflectiveInjector

This change allows ReflectiveInjector to be tree shaken resulting
in not needed Reflect polyfil and smaller bundles.

Code savings for HelloWorld using Closure:

Reflective: bundle.js:  105,864(34,190 gzip)
    Static: bundle.js:  154,889(33,555 gzip)
                            645( 2%)

BREAKING CHANGE:

`platformXXXX()` no longer accepts providers which depend on reflection.
Specifically the method signature when from `Provider[]` to
`StaticProvider[]`.

Example:
Before:
```
[
  MyClass,
  {provide: ClassA, useClass: SubClassA}
]

```

After:
```
[
  {provide: MyClass, deps: [Dep1,...]},
  {provide: ClassA, useClass: SubClassA, deps: [Dep1,...]}
]
```

NOTE: This only applies to platform creation and providers for the JIT
compiler. It does not apply to `@Compotent` or `@NgModule` provides 
declarations.

Benchpress note: Previously Benchpress also supported reflective 
provides, which now require static providers.


DEPRECATION:

- `ReflectiveInjector` is now deprecated as it will be remove. Use
  `Injector.create` as a replacement.

tbosch commented on packages/core/src/di/reflective_provider_interfaces.ts:
```diff
@@ -0,0 +1,95 @@
+ *
+ * @stable
+ */
+export type Provider =
```
`Provider` is also used in our decorators (e.g. `@NgModule`). It seems wrong that it is now defined in a file called `reflective_injector_interfaces`. I think it should be defined in e.g. `di/metadata.ts` and then this file can reexport it if we really need this file.

mhevery replied:
moved back to `provider.ts`

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+  let value: any = EMPTY;
+  let useNew: boolean = false;
+  let provide = resolveForwardRef(provider.provide);
+  if (USE_VALUE in provider) {
```
Why use `in` here? For all the cases we can just check for e.g. `.useFactory !== undefined`, and the fall through case is `useValue`...

tbosch replied:
This would both be faster and remove `getClosureSafeProperty`

mhevery replied:
I can't use `.useValue` since the value could be `undefined`. So since I started using `in` with `useValue`, I am just being consistent.

mhevery replied:
changed, only `USE_VALUE` is special

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+}
+
+interface Records {
+  [key: string]: Record;
```
Why use an object hash and not a `Map` directly?

mhevery replied:
See: https://stackoverflow.com/questions/10258071/class-object-vs-hashmap

Summary: faster, smaller, no polyfill

mhevery replied:
changed to map

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+
+let idCounter = 0;
+
+function getId(obj: any): string {
```
I think using a `Map` is simpler, and shouldn't be slower AFAIK.

mhevery replied:
changed

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+        }
+      }
+      if (useNew) {
+        obj = Object.create(fn.prototype);
```
This won't work if users run in real ES6 mode.
Consider doing a `obj = new fn(...deps)`.

mhevery replied:
good catch, fixed

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+const NG_TOKEN_PATH = 'ngTokenPath';
+const NG_TEMP_TOKEN_PATH = 'ngTempTokenPath';
+const ID_EXPANDO = '__symbol_angular_id_' + new Date().getTime() + '__';
+const OPT_OPTIONAL = 1;
```
Consider using an `enum` for the bitmask here and add the suffix `Flags` (which is what we do everywhere else):
```
enum OptionFlags {
  Optional = 1 << 0,
  CheckSelf = 1 << 1,
  CheckParent = 1 << 2,
}
```

mhevery replied:
Done.

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+    if (!(e instanceof Error)) {
+      e = new Error(e);
+    }
+    const path: any[] = e[NG_TEMP_TOKEN_PATH] = e[NG_TEMP_TOKEN_PATH] || [];
```
Consider passing the `tokenPathOnError` directly to `tryResolveToken`, so you don't need to patch it unto the error (as it is a private field anyways).

mhevery replied:
I don't want to have the cost of `tokenPath` only in the case of an error. Hence it is patched on `Error`

tbosch replied:
I mean you can just create it as an empty array, and only fill it in case of an error in the various `catch` clauses. This way, the cost is only to create the empty array, but not to fill it.

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+      if (depRecords.length) {
+        deps = [];
+        for (let i = 0; i < depRecords.length; i++) {
+          const depRecord: DependencyRecord = depRecords[i], options = depRecord.options,
```
Maybe have separate `const` statements to clear up the formatting?

mhevery replied:
done

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+    } catch (e) {
+      const tokenPath: any[] = e[NG_TEMP_TOKEN_PATH];
+      let message: string = e.message;
+      message = message && message.charAt(0) === NO_NEW_LINE ? message.substr(1) : '\n' + message;
```
Consider adding the logic of adding a newLine or not to `staticError`

mhevery replied:
Done

tbosch commented on packages/core/src/di/injector.ts:
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
+      let message: string = e.message;
+      message = message && message.charAt(0) === NO_NEW_LINE ? message.substr(1) : '\n' + message;
+      const error = staticError(message, tokenPath);
+      e.message = error.message;
```
Did you check if this prints nicely in the browser console? AFAIK it does not.
Another solution would be to throw 2 errors: 1 for flow control where it is now, and another one here with the proper message passed to `new Error(...)`

mhevery replied:
Yes, it looks good in console.

tbosch commented on packages/examples/core/di/ts/provider_spec.ts:
```diff
@@ -89,6 +89,61 @@ export function main() {
 
+    describe('StaticClassProvider', () => {
+      it('works', () => {
+        // #docregion StaticClassProvider
```
I think this is missing a guide that refers to these new `#docregion`s...

mhevery replied:
Not sure I understand. It is present.

tbosch replied:
Ah, never mind, I remembered incorrectly how the guides work.

tbosch reviewed (changes requested):
Looks good in general, needs cleanup.

tbosch commented on packages/benchpress/index.ts:
```diff
@@ -9,7 +9,7 @@
 import 'reflect-metadata';
 
-export {InjectionToken, Injector, Provider, ReflectiveInjector} from '@angular/core';
+export {InjectionToken, Injector, Provider, ReflectiveInjector, StaticProvider} from '@angular/core';
```
Why the changes in benchpress? We might have some code in G3 that relies on this supporting reflective injector...

tbosch replied:
Why the changes in benchpress? We might have some code in G3 that relies on this supporting reflective injector...

tbosch replied:
Never mind, I think it is actually good to change benchpress as well.

tbosch commented:
Type in commit message of the second commit:
```
Specifically the method signature _when_ from `Provider[]` to ...
```

I think the second commit message deserves an example for migration, e.g. a `before` / `after` section.

tbosch commented:
Also for the commit message of the second commit: The providers for the jit compiler, that users were able to also specify, are now `StaticProvider`s.

vicb commented:
blocked on tap/164481065

mhevery commented:
@vicb here is the fix http://cl/164500104

trotyl commented:
Would https://github.com/angular/angular/issues/13820 be affected by this? Seems StaticInjector does not require `@Injectable()` anymore.

sod commented:
`@Injectable()` is just a trick to get typescript to export the typings from the constructor, thanks to `emitDecoratorMetadata` setting in tsconfig.json. It'll most likley stay.

![image](https://user-images.githubusercontent.com/905328/29074110-5bd73cf0-7c4e-11e7-8256-f93c577d034d.png)

trotyl commented:
@sod https://github.com/angular/angular/issues/13820 is about requiring `@Injectable()` when there is no dependency, where metadata is not needed here.

sod commented:
Commit message https://github.com/angular/angular/commit/fcadbf4bf6d00ea5b250a8069e05b3e4bd000a29 reads like you are indeed right. They deprecated or even got rid of the reflective resolver altogether.
